### PR TITLE
feat(SPV-1169) testabilities for record transaction outline

### PR DESCRIPTION
**Pull request by chris-4chain**
This PR waits for #759 

# Pull Request Checklist

- [ ] 📖 I created my PR using provided  : [CODE_STANDARDS](https://github.com/bitcoin-sv/spv-wallet/blob/main/.github/CODE_STANDARDS.md)
- [ ] 📖 I have read the short Code of Conduct: [CODE_OF_CONDUCT](https://github.com/bitcoin-sv/spv-wallet/blob/main/.github/CODE_OF_CONDUCT.md)
- [ ] 🏠 I tested my changes locally.
- [ ] ✅ I have provided tests for my changes.
- [ ] 📝 I have used conventional commits.
- [ ] 📗 I have updated any related documentation.
- [ ] 💾 PR was issued based on the Github or Jira issue.

### CODE REVIEW

File: engine/transaction/record/testabilities/mock_repository.go [hunk 1]
```diff
@@ -1,31 +1,30 @@
...
+func NewMockRepository() *MockRepository {
+	return &MockRepository{
 		transactions: make(map[string]database.Transaction),
 		outputs:      make(map[string]database.Output),
 		data:         make(map[string]database.Data),
 	}
 }
 
-func (m *mockRepository) SaveTX(_ context.Context, txTable *database.Transaction, outputs []*database.Output, data []*database.Data) error {
+func (m *MockRepository) SaveTX(_ context.Context, txTable *database.Transaction, outputs []*database.Output, data []*database.Data) error {
```
Comment (wregulski) [inline engine/transaction/record/testabilities/mock_repository.go:27]: Maybe some testabilities like:
WillFailOnSaveTX as a valuable test case? + such example in `record_outline_test.go`?

File: engine/transaction/record/testabilities/mock_repository.go [hunk 2]
```diff
@@ -36,7 +35,7 @@ func (m *mockRepository) SaveTX(_ context.Context, txTable *database.Transaction
 	return nil
 }
 
-func (m *mockRepository) GetOutputs(_ context.Context, outpoints iter.Seq[bsv.Outpoint]) ([]*database.Output, error) {
+func (m *MockRepository) GetOutputs(_ context.Context, outpoints iter.Seq[bsv.Outpoint]) ([]*database.Output, error) {
```
Comment (wregulski) [inline engine/transaction/record/testabilities/mock_repository.go:38]: And the same proposition here (but maybe it's too deep) to have `WillFailOnGetOutputs`? 🤔

File: engine/transaction/record/record_outline_test.go [hunk 3]
```diff
@@ -278,46 +264,53 @@ func TestRecordOutlineOpReturnErrorCases(t *testing.T) {
...
-
-			// ensure that no changes were made to the repository
-			require.ElementsMatch(t, initialOutputs, test.repo.getAllOutputs())
-			require.ElementsMatch(t, initialData, test.repo.getAllData())
+			then.WithErrorIs(err, test.expectErr).NothingChanged()
 		})
 	}
 }
 
+func TestOnBroadcastErr(t *testing.T) {
```
Comment (wregulski) [inline engine/transaction/record/record_outline_test.go:283]: What do you think about more complex test cases like multiple error handling?
**Example / pseudo code**:
```go
func TestRecordOutlineMultipleFailures(t *testing.T) {
    // given:
    given, then := testabilities.New(t)
    service := given.
        WithStoredOutputs(database.Output{
            TxID:       givenTXWithOpReturn(t).InputUTXO(0).TxID,
            Vout:       givenTXWithOpReturn(t).InputUTXO(0).Vout,
            SpendingTX: nil,
        }).
        WillFailOnSaveTX(errors.New("repository save error")).
        WillFailOnBroadcast(errors.New("broadcaster error")).
        NewRecordService()

    // and:
    outline := &outlines.Transaction{
        BEEF: givenTXWithOpReturn(t).BEEF(),
        Annotations: transaction.Annotations{
            Outputs: transaction.OutputsAnnotations{
                0: &transaction.OutputAnnotation{
                    Bucket: bucket.Data,
                },
            },
        },
    }

    // when:
    err := service.RecordTransactionOutline(context.Background(), outline)

    // then:
    then.WithErrorIs(err, txerrors.ErrTxRepositorySave).NothingChanged()
    then.WithErrorIs(err, txerrors.ErrTxBroadcast).NothingChanged()
}
```

File: engine/transaction/record/testabilities/assert_record_outline.go [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+type ErrorAssert interface {
+	NothingChanged()
+}
+
+type RecordOutlineAssert interface {
+	WithNoError(err error) SuccessfullyCreatedRecordOutlineAssertion
+	WithErrorIs(err, expectedError error) ErrorAssert
+
+	StoredOutputs([]database.Output) RecordOutlineAssert
```
Comment (wregulski) [inline engine/transaction/record/testabilities/assert_record_outline.go:18]: **Just a thought, don't have example 😅**
What do you think about checking if the final "repository" state matches exactly what we intended without any additional entries?

Comment (wregulski) [reply engine/transaction/record/testabilities/assert_record_outline.go:18]: I mean something like:
```go
    then.WithNoError(err).
        Broadcasted(givenTXWithOpReturn(t).ID()).
        StoredOutputs(expectedOutputs).
        StoredData(expectedData)
```

    actualOutputs := given.repository.GetAllOutputs()
    actualData := given.repository.GetAllData()

```go
    require.ElementsMatch(t, expectedOutputs, actualOutputs, "Repository contains unexpected outputs")
    require.ElementsMatch(t, expectedData, actualData, "Repository contains unexpected data")
```

And at the top define some expectedData used here as a parameter for require 😃

File: engine/transaction/record/testabilities/mock_repository.go (hunk 1 above)
Comment (chris-4chain) [reply engine/transaction/record/testabilities/mock_repository.go:27]: Good idea 
I'm adding this `WillFailOnSaveTX ` and a test case for that 🚀

File: engine/transaction/record/record_outline_test.go (hunk 3 above)
Comment (chris-4chain) [reply engine/transaction/record/record_outline_test.go:283]: That won't work like you're showing.
The error can be either `ErrTxBroadcast` or `ErrTxRepositorySave` - but not both at the same time.
Taking this into account - I don't think, such test will be beneficial.

File: engine/transaction/record/testabilities/assert_record_outline.go (hunk 4 above)
Comment (chris-4chain) [reply engine/transaction/record/testabilities/assert_record_outline.go:18]: I've changed from `Subset` to `ElementsMatch`, so now the behawior is as you described.

File: engine/transaction/record/testabilities/mock_repository.go (hunk 2 above)
Comment (chris-4chain) [reply engine/transaction/record/testabilities/mock_repository.go:38]: No problem 👍

File: engine/transaction/record/testabilities/mock_repository.go (hunk 1 above)
Comment (chris-4chain) [reply engine/transaction/record/testabilities/mock_repository.go:30]: Done 😄

File: engine/transaction/record/testabilities/assert_record_outline.go (hunk 4 above)
Comment (chris-4chain) [reply engine/transaction/record/testabilities/assert_record_outline.go:18]: Done 😄

File: engine/transaction/record/record_outline_test.go (hunk 3 above)
Comment (wregulski) [reply engine/transaction/record/record_outline_test.go:278]: Ok, undesrtand 👍

**Review by wregulski — APPROVED** [review]

File: engine/transaction/record/record_outline_test.go [hunk 5]
```diff
@@ -278,46 +259,78 @@ func TestRecordOutlineOpReturnErrorCases(t *testing.T) {
...
 			err := service.RecordTransactionOutline(context.Background(), test.outline)
 
 			// then:
-			require.Error(t, err)
-			require.ErrorIs(t, err, test.expectErr)
-
-			// ensure that no changes were made to the repository
-			require.ElementsMatch(t, initialOutputs, test.repo.getAllOutputs())
-			require.ElementsMatch(t, initialData, test.repo.getAllData())
+			then.WithErrorIs(err, test.expectErr).NothingChanged()
```
Comment (dorzepowski) [inline engine/transaction/record/record_outline_test.go:273]: Maybe?
```suggestion
			then.ErrorIs(err, test.expectErr).
			NothingChangedInRepository()
```

File: engine/transaction/record/record_outline_test.go [hunk 6]
```diff
@@ -278,46 +259,78 @@ func TestRecordOutlineOpReturnErrorCases(t *testing.T) {
...
+	// given:
+	given, then := testabilities.New(t)
+	service := given.
+		WithStoredOutputs(database.Output{
+			TxID:       givenTXWithOpReturn(t).InputUTXO(0).TxID,
+			Vout:       givenTXWithOpReturn(t).InputUTXO(0).Vout,
+			SpendingTX: nil,
+		}).
+		WillFailOnBroadcast(errors.New("broadcast error")).
+		NewRecordService()
```
Comment (dorzepowski) [inline engine/transaction/record/record_outline_test.go:288]: I would rather add context to calls, to increase understandability what is the environment setup

```suggestion
	given.Repository().HasOutputs(database.Output{
			TxID:       givenTXWithOpReturn(t).InputUTXO(0).TxID,
			Vout:       givenTXWithOpReturn(t).InputUTXO(0).Vout,
			SpendingTX: nil,
		})
	given.Broadcaster().WillFailOnBroadcast(errors.New("broadcast error")).
		
	service := given.NewRecordService()
```



or:

```suggestion
	given.Repository().HasOutputs(database.Output{
			TxID:       givenTXWithOpReturn(t).InputUTXO(0).TxID,
			Vout:       givenTXWithOpReturn(t).InputUTXO(0).Vout,
			SpendingTX: nil,
		}).
	And().Broadcaster().WillFailOnBroadcast(errors.New("broadcast error")).
		
	service := given.NewRecordService()
```

File: engine/transaction/record/record_outline_test.go [hunk 7]
```diff
@@ -278,46 +259,78 @@ func TestRecordOutlineOpReturnErrorCases(t *testing.T) {
...
-		},
 	}
 	for name, test := range tests {
 		t.Run(name, func(t *testing.T) {
 			// given:
-			service := record.NewService(tester.Logger(t), test.repo, test.broadcaster)
-			initialOutputs := test.repo.getAllOutputs()
-			initialData := test.repo.getAllData()
+			given, then := testabilities.New(t)
+			service := given.WithStoredOutputs(test.storedOutputs...).NewRecordService()
```
Comment (dorzepowski) [inline engine/transaction/record/record_outline_test.go:267]: I would rather make it separated calls like this:
```suggestion
			given.WithStoredOutputs(test.storedOutputs...)
			
			service := given.NewRecordService()
```

File: engine/transaction/record/testabilities/mock_repository.go [hunk 8]
```diff
@@ -1,31 +1,36 @@
...
 	"maps"
 	"slices"
 
-	trx "github.com/bitcoin-sv/go-sdk/transaction"
 	"github.com/bitcoin-sv/spv-wallet/engine/database"
 	"github.com/bitcoin-sv/spv-wallet/models/bsv"
 )
 
-type mockRepository struct {
+type MockRepository struct {
```
Comment (dorzepowski) [inline engine/transaction/record/testabilities/mock_repository.go:13]: maybe it doesn't need to be public yet?

File: engine/transaction/record/testabilities/mock_repository.go [hunk 9]
```diff
@@ -1,31 +1,36 @@
...
 	outputs      map[string]database.Output
 	data         map[string]database.Data
+
+	errOnSave       error
+	errOnGetOutputs error
 }
 
-func newMockRepository() *mockRepository {
-	return &mockRepository{
+func NewMockRepository() *MockRepository {
```
Comment (dorzepowski) [inline engine/transaction/record/testabilities/mock_repository.go:22]: maybe it doesn't need to be public yet?

File: engine/transaction/record/testabilities/mock_repository.go [hunk 10]
```diff
@@ -46,44 +54,41 @@ func (m *mockRepository) GetOutputs(_ context.Context, outpoints iter.Seq[bsv.Ou
 	return outputs, nil
 }
 
-func (m *mockRepository) withOutput(output database.Output) *mockRepository {
+func (m *MockRepository) WithOutput(output database.Output) *MockRepository {
```
Comment (dorzepowski) [inline engine/transaction/record/testabilities/mock_repository.go:57]: ### suggestion

`AddOutput` or `Add`

File: engine/transaction/record/testabilities/assert_record_outline.go [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+	require.Equal(a.t, database.TxStatusBroadcasted, tx.TxStatus, "Transaction %s is not stored as broadcasted", txID)
+	return a
+}
+
+func (a *assert) StoredOutputs(outputs []database.Output) RecordOutlineAssert {
+	require.ElementsMatch(a.t, a.given.repository.GetAllOutputs(), outputs)
+	return a
+}
+
+func (a *assert) StoredData(data []database.Data) RecordOutlineAssert {
```
Comment (dorzepowski) [inline engine/transaction/record/testabilities/assert_record_outline.go:71]: ### minor 
And we can get back to that later,

BUT maybe `Data` is too ambiguous word, maybe at least we should add some prefix like OutputData or something :shrug:

**Review by dorzepowski — CHANGES_REQUESTED** [review]
Overall great job ❤️ 
I just add some points (comments) to increase understandability what is the state when calling the methods.

File: engine/transaction/record/record_outline_test.go (hunk 5 above)
Comment (chris-4chain) [reply engine/transaction/record/record_outline_test.go:273]: Sure :)

File: engine/transaction/record/record_outline_test.go (hunk 6 above)
Comment (chris-4chain) [reply engine/transaction/record/record_outline_test.go:288]: I added `Repository` and `Broadcaster` methods - so the `given` test configuration is setup without `RecordServiceFixture` as "proxy"

File: engine/transaction/record/record_outline_test.go (hunk 7 above)
Comment (chris-4chain) [reply engine/transaction/record/record_outline_test.go:267]: Adjusted 👍

File: engine/transaction/record/testabilities/mock_repository.go (hunk 8 above)
Comment (chris-4chain) [reply engine/transaction/record/testabilities/mock_repository.go:13]: After my changes (inspired by your comments ☝️) - this requires to be exported (also MockBroadcaster)

File: engine/transaction/record/testabilities/mock_repository.go (hunk 10 above)
Comment (chris-4chain) [reply engine/transaction/record/testabilities/mock_repository.go:57]: I like `WithOutput` name.

File: engine/transaction/record/testabilities/assert_record_outline.go (hunk 11 above)
Comment (chris-4chain) [reply engine/transaction/record/testabilities/assert_record_outline.go:71]: Let's get back to this topic later. 🙏

File: engine/transaction/record/testabilities/mock_repository.go (hunk 9 above)
Comment (chris-4chain) [reply engine/transaction/record/testabilities/mock_repository.go:22]: sure 👍

**Review by dorzepowski — APPROVED** [review]

**Pull request merged**